File: test/oracle.py
# Independent oracle for the thymus example, sharing no code with the SQL pipeline:
# read the GFF3 by hand, build every category interval, and for each peak centre scan
# all intervals for the best priority. Slow on purpose. Writes
# test/fixtures/thymus-expected.json. Run from the repository root: python3 test/oracle.py
import json
import gzip, bisect, collections, urllib.parse
from pathlib import Path
UP = DOWN = 1000
CAT = ["promoter","utr5","utr3","exon","intron","intergenic"]
rows = [l.rstrip("\n").split("\t") for l in gzip.open("examples/gencode.vM25.basic.chr19.gff3.gz","rt") if not l.startswith("#")]
def attrs(s): return {k: urllib.parse.unquote(v) for k,_,v in (p.partition("=") for p in s.split(";") if "=" in p)}
byid = {}; exon_parents = set(); parts = []
for c in rows:
    a = attrs(c[8]); s, e = int(c[3]) - 1, int(c[4])
    if "ID" in a: byid.setdefault(a["ID"], (c[0], s, e, c[6]))
    t = c[2].lower()
    for p in a.get("Parent","").split(","):
        if p and t in ("exon","five_prime_utr","three_prime_utr"):
            parts.append((p, t, c[0], s, e))
            if t == "exon": exon_parents.add(p)
iv = collections.defaultdict(list)  # chrom -> (s,e,prio)
for tx in exon_parents:
    chrom, s, e, strand = byid[tx]
    t = s if strand == "+" else e - 1
    ps, pe = (t-UP, t+DOWN+1) if strand == "+" else (t-DOWN, t+UP+1)
    iv[chrom].append((max(0,ps), pe, 1)); iv[chrom].append((s, e, 5))
for tx, t, chrom, s, e in parts:
    if tx in exon_parents:
        iv[chrom].append((s, e, {"five_prime_utr":2,"three_prime_utr":3,"exon":4}[t]))
def cat(chrom, x):
    best = 6
    for s, e, p in iv.get(chrom, ()):
        if s <= x < e and p < best: best = p
    return best
out = {}
for f in sorted(Path("examples").glob("*.narrowPeak.gz")):
    n = collections.Counter()
    for l in gzip.open(f, "rt"):
        c = l.split("\t"); s, e = int(c[1]), int(c[2])
        n[CAT[cat(c[0], s + (e - s)//2) - 1]] += 1
    out[f.name.split("_")[1]] = {k: n[k] for k in CAT}
Path("test/fixtures/thymus-expected.json").write_text(json.dumps(
    {"_comment": "Peak-centre counts from test/oracle.py, promoter 1000/1000, all transcripts.", "counts_centre": out},
    indent=2) + "\n")
print(json.dumps(out))

File: test/oracle.R
# Independent oracle for the thymus data. It shares no code with the SQL pipeline
# and uses base R only: read the GFF3 by hand, build every category interval, and give
# each peak centre the best (lowest) priority among the intervals that contain it.
# Slow on purpose. Defaults to the bundled chr19 data and expected fixture.
#
# Run from the repository root: Rscript test/oracle.R [annotation.gff3.gz peak-dir output.json]

args <- commandArgs(TRUE)
stopifnot(length(args) %in% c(0L, 3L))
annotation <- if (length(args) == 0L) "examples/gencode.vM25.basic.chr19.gff3.gz" else args[1]
peak_directory <- if (length(args) == 0L) "examples" else args[2]
output <- if (length(args) == 0L) "test/fixtures/thymus-expected.json" else args[3]

promoter_upstream <- 1000L
promoter_downstream <- 1000L
categories <- c("promoter", "utr5", "utr3", "exon", "intron", "intergenic")

read_gz_lines <- function(path) {
  con <- gzfile(path, "rt")
  on.exit(close(con))
  readLines(con)
}

gff_attributes <- function(column) {
  lapply(strsplit(column, ";", fixed = TRUE), function(pairs) {
    pairs <- pairs[grepl("=", pairs, fixed = TRUE)]
    keys <- sub("=.*$", "", pairs)
    values <- vapply(sub("^[^=]*=", "", pairs), URLdecode, character(1), USE.NAMES = FALSE)
    setNames(values, keys)
  })
}

lines <- read_gz_lines(annotation)
lines <- lines[!startsWith(lines, "#")]
fields <- do.call(rbind, strsplit(lines, "\t", fixed = TRUE))
gff <- data.frame(
  chrom = fields[, 1],
  type = tolower(fields[, 3]),
  s = as.integer(fields[, 4]) - 1L, # 0-based half-open from here on
  e = as.integer(fields[, 5]),
  strand = fields[, 7],
  stringsAsFactors = FALSE
)
attrs <- gff_attributes(fields[, 9])
gff$id <- vapply(attrs, function(a) if ("ID" %in% names(a)) a[["ID"]] else NA_character_, "")
parents <- lapply(attrs, function(a) if ("Parent" %in% names(a)) strsplit(a[["Parent"]], ",")[[1]] else character())

# One row per (parent, part) for exons and UTRs.
part_types <- c("exon", "five_prime_utr", "three_prime_utr")
part_rows <- which(gff$type %in% part_types)
parts <- do.call(rbind, lapply(part_rows, function(i) {
  if (length(parents[[i]]) == 0L) return(NULL)
  data.frame(tx = parents[[i]], type = gff$type[i], chrom = gff$chrom[i],
             s = gff$s[i], e = gff$e[i], stringsAsFactors = FALSE)
}))

# A transcript is any feature that is the Parent of an exon; the first feature with a
# given ID supplies its span and strand.
transcripts <- unique(parts$tx[parts$type == "exon"])
by_id <- gff[!is.na(gff$id) & !duplicated(gff$id), ]
tx <- by_id[match(transcripts, by_id$id), ]

tss <- ifelse(tx$strand == "+", tx$s, tx$e - 1L)
promoter_s <- ifelse(tx$strand == "+", tss - promoter_upstream, tss - promoter_downstream)
promoter_e <- ifelse(tx$strand == "+", tss + promoter_downstream + 1L, tss + promoter_upstream + 1L)
part_priority <- c(five_prime_utr = 2L, three_prime_utr = 3L, exon = 4L)
kept_parts <- parts[parts$tx %in% transcripts, ]

intervals <- rbind(
  data.frame(chrom = tx$chrom, s = pmax(0L, promoter_s), e = promoter_e, priority = 1L),
  data.frame(chrom = tx$chrom, s = tx$s, e = tx$e, priority = 5L),
  data.frame(chrom = kept_parts$chrom, s = kept_parts$s, e = kept_parts$e,
             priority = unname(part_priority[kept_parts$type]))
)
intervals_by_chrom <- split(intervals, intervals$chrom)

category_of <- function(chrom, x) {
  iv <- intervals_by_chrom[[chrom]]
  if (is.null(iv)) return(6L)
  hit <- iv$priority[iv$s <= x & x < iv$e]
  if (length(hit) == 0L) 6L else min(hit)
}

peak_files <- sort(list.files(peak_directory, pattern = "\\.narrowPeak\\.gz$", full.names = TRUE))
counts <- lapply(peak_files, function(path) {
  peak <- do.call(rbind, strsplit(read_gz_lines(path), "\t", fixed = TRUE))
  s <- as.integer(peak[, 2])
  e <- as.integer(peak[, 3])
  centre <- s + (e - s) %/% 2L
  category <- mapply(category_of, peak[, 1], centre, USE.NAMES = FALSE)
  setNames(tabulate(category, nbins = length(categories)), categories)
})
names(counts) <- vapply(strsplit(basename(peak_files), "_", fixed = TRUE), `[`, "", 2L)

json_counts <- function(n, indent) {
  body <- paste0(indent, '  "', names(n), '": ', n, collapse = ",\n")
  paste0("{\n", body, "\n", indent, "}")
}
json <- paste0(
  "{\n",
  '  "_comment": "Peak-centre counts from test/oracle.R, promoter 1000/1000, all transcripts.",\n',
  '  "counts_centre": {\n',
  paste0('    "', names(counts), '": ', vapply(counts, json_counts, "", indent = "    "), collapse = ",\n"),
  "\n  }\n}\n"
)
writeLines(json, output, sep = "")
invisible(lapply(names(counts), function(n) cat(n, counts[[n]], "\n")))
